File: Python/Recover Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, modify the binary tree in-place instead.
    def recoverTree(self, root):
        if(root==None):return None
        else:
            stack=[]
            self.genStack(root,stack)
            vals= map(lambda x:x.val,stack)
            vals=sorted(vals)
    
            for i in range(len(stack)):
                stack[i].val=vals[i]
    
    def genStack(self, node,stack):            
        if(node==None):return
        self.genStack(node.left,stack)
        stack.append(node)
        self.genStack(node.right,stack)

File: Java/src/Recover Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    TreeNode first = null;
    TreeNode sec = null;
    TreeNode prev = new TreeNode(Integer.MIN_VALUE);

    public void recoverTree(TreeNode root) {

        helper(root);

        int temp = first.val;
        first.val = sec.val;
        sec.val = temp;
    }

    private void helper(TreeNode root) {
        if (root == null)
            return;

        helper(root.left);

        if (first == null && prev.val >= root.val) {
            first = prev;
        }

        if (first != null && prev.val >= root.val) {
            sec = root;
        }
        
        prev = root;

        helper(root.right);
}
}